src/pygeon/discretizations/fem/h1.py

```diff
@@ -347,3 +350,551 @@ def get_range_discr_class(self, dim: int) -> pg.Discretization:
+
+        # Helper constants
+        n_edges = self.num_edges_per_cell(dim)
+        eye = np.eye(dim + 1)
+        zero = np.zeros((n_edges, dim + 1))
+
+        # List the barycentric functions up to degree 2,
+        # by exponents, consisting of
+        # - the linears lambda_i
+        # - the cross-quadratics lambda_i lambda_j
+        # - the quadratics lambda_i^2
+        quads = np.zeros((dim + 1, n_edges))
+        e_nodes = self.get_local_edge_nodes(dim)
+        for ind, nodes in enumerate(e_nodes):
+            quads[nodes, ind] = 1
+        exponents = np.hstack((eye, quads, 2 * eye))
+
+        # Compute the local mass matrix of the barycentric functions
+        barycentric_mass = self.assemble_barycentric_mass(exponents)
+
+        # Our basis functions are given by
+        # - nodes: lambda_i (2 lambda_i - 1)
+        # - edges: 4 lambda_i lambda_j
+        # We list the coefficients in the array "basis"
+        basis_nodes = np.vstack((-eye, zero, 2 * eye))
+        basis_edges = np.zeros((2 * (dim + 1) + n_edges, n_edges))
+        basis_edges[dim + 1 : dim + n_edges + 1, :] = 4 * np.eye(n_edges)
+        basis = np.hstack((basis_nodes, basis_edges))
+
+        return basis.T @ barycentric_mass @ basis
+
+    def assemble_barycentric_mass(self, expnts: np.ndarray) -> np.ndarray:
+        """
+        Compute the inner products of all monomials up to degree 2
+
+        Args:
+            expnts (np.ndarray): each column is an array of exponents
+                alpha_i of the monomial expressed as
+                prod_i lambda_i ^ alpha_i.
+
+        Returns:
+            np.ndarray: the inner products of the monomials
+                on a simplex with measure 1.
+        """
+        n_monomials = expnts.shape[1]
+        mass = np.empty((n_monomials, n_monomials))
+
+        for i in np.arange(n_monomials):
+            for j in np.arange(n_monomials):
+                mass[i, j] = self.integrate_monomial(expnts[:, i] + expnts[:, j])
+
+        return mass
+
+    @staticmethod
+    def factorial(n: float) -> int:
+        """
+        Compute the factorial of a float by first rounding to an int.
+        Args:
+            n (float): the input float
+
+        Returns:
+            int: the factorial n!
+        """
+        return math.factorial(int(n))
+
+    def integrate_monomial(self, alphas: np.ndarray) -> float:
+        """
+        Exact integration of products of monomials based on
+        Vermolen and Segal (2018).
+
+        Args:
+            alphas (np.ndarray): array of exponents alpha_i of the monomial
+                expressed as prod_i lambda_i ^ alpha_i
+
+        Returns:
+            float: the integral of the monomial on a simplex with measure 1
+        """
+        dim = len(alphas) - 1
+        fac_alph = [self.factorial(a_i) for a_i in alphas]
+
+        return (
+            self.factorial(dim)
+            * np.prod(fac_alph)
+            / self.factorial(dim + np.sum(alphas))
```

Comment:
if the creation of the self.factorial is needed only here, we can use np.sum(alpha, dtype=int) instead